lib/spack/spack/fetch_strategy.py

```diff
@@ -337,26 +355,37 @@ def expand(self):
         with working_dir(tarball_container):
             decompress(self.archive_file)
 
-        # Check for an exploding tarball, i.e. one that doesn't expand
-        # to a single directory.  If the tarball *didn't* explode,
-        # move contents up & remove the container directory.
+        # Check for an exploding tarball, i.e. one that doesn't expand to
+        # a single directory.  If the tarball *didn't* explode, move its
+        # contents to the staging source directory & remove the container
+        # directory.  If the tarball did explode, just rename the tarball
+        # directory to the staging source directory.
         #
-        # NOTE: The tar program on Mac OS X will encode HFS metadata
-        # in hidden files, which can end up *alongside* a single
-        # top-level directory.  We ignore hidden files to accomodate
-        # these "semi-exploding" tarballs.
+        # NOTE: The tar program on Mac OS X will encode HFS metadata in
+        # hidden files, which can end up *alongside* a single top-level
+        # directory.  We initially ignore presence of hidden files to
+        # accomodate these "semi-exploding" tarballs but ensure the files
+        # are copied to the source directory.
         files = os.listdir(tarball_container)
         non_hidden = [f for f in files if not f.startswith('.')]
         if len(non_hidden) == 1:
-            expanded_dir = os.path.join(tarball_container, non_hidden[0])
-            if os.path.isdir(expanded_dir):
-                for f in files:
-                    shutil.move(os.path.join(tarball_container, f),
-                                os.path.join(self.stage.path, f))
+            src = os.path.join(tarball_container, non_hidden[0])
+            if os.path.isdir(src):
+                shutil.move(src, self.stage.source_path)
+                if len(files) > 1:
+                    files.remove(non_hidden[0])
+                    for f in files:
+                        src = os.path.join(tarball_container, f)
+                        dest = os.path.join(self.stage.path, f)
+                        shutil.move(src, dest)
```

Comment:
I suspect my issue #11645 is coming from these lines.  Like the tarball is not being unpacked where it is expected to be?  Just a guess.

This is also affecting other packages, example:

```
[osolberg@bifx1n03 spack]$ ./bin/spack install lua-luafilesystem
==> Installing libsigsegv
==> Searching for binary cache of libsigsegv
==> Warning: No Spack mirrors are currently configured
==> No binary for libsigsegv found: installing from source
==> Fetching https://ftpmirror.gnu.org/libsigsegv/libsigsegv-2.11.tar.gz
######################################################################## 100.0%
==> Staging archive: /home/osolberg/temp/spack/var/spack/stage/libsigsegv-2.11-qpmaxx6z62df4s4hwyehdwptv6kmrfhf/libsigsegv-2.11.tar.gz
==> Created stage in /home/osolberg/temp/spack/var/spack/stage/libsigsegv-2.11-qpmaxx6z62df4s4hwyehdwptv6kmrfhf
==> No patches needed for libsigsegv
==> Building libsigsegv [AutotoolsPackage]
==> Error: OSError: [Errno 2] No such file or directory: '/home/osolberg/temp/spack/var/spack/stage/libsigsegv-2.11-qpmaxx6z62df4s4hwyehdwptv6kmrfhf/src'

/home/osolberg/temp/spack/lib/spack/spack/package.py:1577, in build_process:
       1574                    echo = logger.echo
       1575                    self.log()
       1576
  >>   1577                # Run post install hooks before build stage is removed.
       1578                spack.hooks.post_install(self.spec)
       1579
       1580            # Stop timer.
```

But `src` isn't at `var/spack/stage/libsigsegv-2.11-qpmaxx6z62df4s4hwyehdwptv6kmrfhf`, it's in `var/spack/stage/libsigsegv-2.11-qpmaxx6z62df4s4hwyehdwptv6kmrfhf/libsigsegv-2.11`:

```
[spack]$ ls -l var/spack/stage/libsigsegv-2.11-qpmaxx6z62df4s4hwyehdwptv6kmrfhf/
total 444
drwxr-xr-x 6 osolberg domain users   4096 Feb 18  2017 libsigsegv-2.11
-rw-r--r-- 1 osolberg domain users 448744 Jun  6 09:54 libsigsegv-2.11.tar.gz
[spack]$ ls -l var/spack/stage/libsigsegv-2.11-qpmaxx6z62df4s4hwyehdwptv6kmrfhf/libsigsegv-2.11/
total 828
-rw-r--r-- 1 osolberg domain users  43625 Feb 18  2017 aclocal.m4
-rw-r--r-- 1 osolberg domain users    242 Jan  2  2014 AUTHORS
drwxr-xr-x 2 osolberg domain users    128 Feb 18  2017 build-aux
-rw-r--r-- 1 osolberg domain users  49040 Feb 18  2017 ChangeLog
-rw-r--r-- 1 osolberg domain users  12345 Jan  2  2014 ChangeLog.1
-rw-r--r-- 1 osolberg domain users   4564 Feb 18  2017 config.h.in
-rw-r--r-- 1 osolberg domain users   4583 Feb 18  2017 config.h.msvc
-rwxr-xr-x 1 osolberg domain users 552750 Feb 18  2017 configure
-rw-r--r-- 1 osolberg domain users  31050 Feb 18  2017 configure.ac
-rw-r--r-- 1 osolberg domain users  18154 Jan  2  2014 COPYING
-rw-r--r-- 1 osolberg domain users   9678 Jan  2  2014 INSTALL
drwxr-xr-x 2 osolberg domain users    331 Feb 18  2017 m4
-rw-r--r-- 1 osolberg domain users   2830 Nov 23  2016 Makefile.am
-rw-r--r-- 1 osolberg domain users  28269 Feb 18  2017 Makefile.in
-rw-r--r-- 1 osolberg domain users   3385 Jan  2  2014 Makefile.msvc
-rw-r--r-- 1 osolberg domain users   4489 Feb 18  2017 NEWS
-rw-r--r-- 1 osolberg domain users  23457 Feb 18  2017 PORTING
-rw-r--r-- 1 osolberg domain users   6229 Feb 18  2017 README
-rw-r--r-- 1 osolberg domain users  10367 Nov 23  2016 README.windows
drwxr-xr-x 2 osolberg domain users   4096 Feb 18  2017 src
drwxr-xr-x 2 osolberg domain users    250 Feb 18  2017 tests
```